signalfx-java/src/main/java/com/signalfx/signalflow/WebSocketTransport.java

```diff
@@ -0,0 +1,592 @@
+import org.eclipse.jetty.websocket.WebSocketClient;
+import org.eclipse.jetty.websocket.WebSocketClientFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.DeserializationFeature;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableMap;
+import com.signalfx.endpoint.SignalFxEndpoint;
+import com.signalfx.signalflow.ChannelMessage.Type;
+
+/**
+ * WebSocket based transport.
+ * 
+ * Uses the SignalFlow WebSocket connection endpoint to interact with SignalFx's SignalFlow API.
+ * Multiple computation streams can be multiplexed through a single, pre-opened WebSocket
+ * connection. It also utilizes a more efficient binary encoding for data so it requires less
+ * bandwidth and has overall less latency.
+ * 
+ * @author dgriff
+ */
+public class WebSocketTransport implements SignalFlowTransport {
+
+    protected static final Logger log = LoggerFactory.getLogger(WebSocketTransport.class);
+    public static final Integer DEFAULT_TIMEOUT = 1; // 1 second
+
+    protected final String token;
+    protected final SignalFxEndpoint endpoint;
+    protected final String path;
+    protected Integer timeout = DEFAULT_TIMEOUT;
+    protected WebSocketClient webSocketClient;
+    protected TransportConnection transportConnection;
+
+    protected WebSocketTransport(String token, final SignalFxEndpoint endpoint,
+                                 final int apiVersion, final Integer timeout) {
+        this.token = token;
+        this.endpoint = endpoint;
+        this.path = "/v" + apiVersion + "/signalflow/connect";
+        this.timeout = timeout;
+
+        try {
+            WebSocketClientFactory factory = new WebSocketClientFactory();
+            factory.start();
+
+            this.webSocketClient = factory.newWebSocketClient();
+
+            URIBuilder uriBuilder = new URIBuilder(String.format("%s://%s:%s%s",
+                    endpoint.getScheme(), endpoint.getHostname(), endpoint.getPort(), path));
+
+            this.transportConnection = new TransportConnection(token);
+
+            this.webSocketClient.open(uriBuilder.build(), this.transportConnection, timeout,
+                    TimeUnit.SECONDS);
+
+        } catch (Exception ex) {
+            throw new SignalFlowException("failed to construct websocket transport", ex);
+        }
+    }
+
+    @Override
+    public Channel attach(String handle, final Map<String, String> parameters) {
+        if (log.isDebugEnabled()) {
+            log.debug("attach: [ " + handle + " ] with parameters: " + parameters);
+        }
+
+        Channel channel = new TransportChannel(transportConnection);
+        HashMap<String, String> request = new HashMap<String, String>(parameters);
+        request.putAll(buildParams("type", "attach", "handle", handle));
```

Comment:
If you're adding string/string pairs, avoid building the additional map that `buildParams` builds:

```java
Map<String, String> request = new HashMap<>();
request.putAll(parameters);
request.put("type", "attach");
request.put("handle", handle);
```